src/com/carolinarollergirls/scoreboard/model/SkaterModel.java

```diff
@@ -17,17 +17,34 @@
 	public Skater getSkater();
 
 	public void bench();
-	public void unBench();
+	public void restoreSnapshot(SkaterSnapshot s);
 
 	public void setName(String id);
 	public void setNumber(String number);
 	public void setPosition(String position) throws PositionNotFoundException;
 	public void setPenaltyBox(boolean box);
 	public void setFlags(String flags);
 
-  // A null code removes the penalty.  
-  public void AddPenaltyModel(String id, boolean foulout_explusion, int period, int jam, String code);
+	// A null code removes the penalty.  
+	public void AddPenaltyModel(String id, boolean foulout_explusion, int period, int jam, String code);
 
 	public static interface PenaltyModel extends Penalty {
 	}
+	
+	public class SkaterSnapshot	{
+		public SkaterSnapshot(SkaterModel skater) {
```

Comment:
Models shouldn't contain any code, they're just interfaces so should only have methods.

Reply:
There's no need to have the implementation in the interface - and none of the other views or models do this. It's also fine to have an inner class that's not mentioned in the Model, if it's not doing to be part of the Model interface.

Reply:
Okay, then keep it in the model but remove the implementation from the model. The implementation should go in the default class.

Reply:
I just wrote that I'd rather not do this and gave a reason. Now just demanding I do it without addressing my reasoning at all is not going to bring us towards a consensus.

Reply:
I got that first part. My point wass that duplicating 70% of the class in the interface is ugly and annoying to maintain. I didn't consider the option of having an Interface that doesn't contain the actual interface. 
So I'll implement the latter when I get home, though I don't consider it a good way of doing it but just less bad than the other options.

Reply:
I understand this, there are many aspects of this codebase that are less than optimal. I say we work with what we have and try to keep things consistent, there's already enough sedimentary layers in this codebase that we don't need to add more.

Reply:
...except that a methodless interface won't work because then the implementation of `restoreSnapshot`, which also gets the interface as parameter, won't be able to access the getter methods (unless we cast to the assumed implementation, which violates type safety).

So I'll revert to my initial implementation of having the Snapshot classes separate. While this means adding a new parameter that should be part of the snapshot requires editing two classes across two files, the other option would require editing two classes and an interface across two files, which is worse.

Reply:
It's not a problem to have mehods on the interface. Penalty has methods, but they're in the view interface rather than the model and you can do the same here.

Once again, whether they are inner classes or classes in their own right is completely orthogonal.

Reply:
It's not a problem in the sense that it wouldn't work or violate any design principles. But it would give us the downsides that originally caused the request to have the snapshots as inner classes. So there's no point in having them as inner classes.

Reply:
Even as standalone classes, the current design pattern would still require a view interface and a model interface for those classes.

The snapshots are quite tightly bound the the things their snapshotting in, so it feels more natural to me to have them as inner classes. It'd be weird if Penalty and TeamColor is an inner class and Snapshot isn't.

Reply:
As we have seen here, having them as inner classes would require duplicating ~70% of the code or storing a single snapshot distributed over dozens of objects. So the approach currently in this PR is the least bad IMO.
(And the View-Model-Implementation pattern isn't consistently followed anyway.)

Reply:
I'm not following you, how does them being inner classes affect code size?

> (And the View-Model-Implementation pattern isn't consistently followed anyway.)

Where is it not being followed? I'll have to add it to my todo list.

Reply:
> how does them being inner classes affect code size?

Ok, once more: We can't have the snapshot as an inner class without declaring its full interface in the corresponding Model. Given the nature of the snapshot classes consisting of just a construcotr and a set of getters, these interfaces will then repeat a large part of the code of the actual class.

> Where is it not being followed? I'll have to add it to my todo list.

- Ruleset and rules
- The event system
- The jetty, json, viewer, and xml packages
- The metrics collector
- The policy package (huh, I thought I removed that o.o)
- ...

Good luck with your todo list.

Reply:
> We can't have the snapshot as an inner class without declaring its full interface in the corresponding Model.

Ah, that's what you're talking about. That's normal for this design pattern, it's just something you have to do. I'm surprised Eclipse doesn't have tooling to handle it for you, though I've always done it by hand.

> Good luck with your todo list.

The MVC model is only followed within the ScoreBoard class and its components, as it is the M and V. The rest is the C around the MV. Of those you list only rulesets needs doing, which I've already on my todo list. Teams also need it.

Reply:
But if you view the scoreboard as MVC, the snapshots are not part of the outward face of the model. They are internal to the model (though not to a single class of the model). So they don't need to be part of the model interface.

Reply:
> So they don't need to be part of the model interface.

If you can make it work with them not being exposed as part of the X/XModel I'm cool with that, though I'm not sure that's practical.

I propose that you go back to the approach in https://github.com/rollerderby/scoreboard/pull/125/commits/1f549869e758f7683a7dcb1feede2fbdd4a0de2a, but with the implementation of all those methods living to the DefafultXModel.java rather than in X or XModel.java consistent with the existing codebase.

One of the things I'm currently slowly working on with the scoreboard is getting all state to live under ScoreBoard somewhere, getting it into the WS, and very long term getting rid of XML and all that manual handling. Penalties are done and Stats is currently what I'm working on.
These snapshots are state, so they should live with all the other state as they are both used by and use the current state classes. The circular dependency in this snapshot package is also a bit of a smell, these aren't pure utility classes like say ClockConversion.

Reply:
What you want is one package that contains all scoreboard state and has a coherent outward facing interface. But that's a long way from what we have today, with state split across several packages (defalt, model, top level, ...) and each class having a seperate Interface that mixes outward and inward facing methods. Unless we attack that underlying problem there won't be a good solution for the snapshots. And which of the bad solutions is least bad is a matter of taste more than anything else. I don't see why your taste should be more relevant than mine here.

And BTW: This discussion is blocking a possible fix for the random clock weirdnesses and a fix for #48, which is marked as urgent, while we are less than 2 weeks from the first continental cup.

Reply:
> What you want is one package that contains all scoreboard state and has a coherent outward facing interface. But that's a long way from what we have today, with state split across several packages (defalt, model, top level, ...) and each class having a seperate Interface that mixes outward and inward facing methods

More precisely, I want all our data under the ScoreBoard interface and following its design patterns. I don't want data living only in the XML (e.g. keybindings) or other ad-hoc designs (rulesets). Once that massive task is done, we can get rid of XML and more importantly all the manual handling that goes with it.

I don't see a reason to change from the default/model/top MVC pattern, and in fact plan to move a lot more of CRG to it. Sure it's a little bit verbose, but that's a price worth paying to have a coherent, clean, and maintainable design.

> Unless we attack that underlying problem there won't be a good solution for the snapshots. 

I am actively attacking the problem. I've done it with penalties, I'm at the tail end of doing it with stats, and after that I'm planning on doing rulesets. Then it's all the state that lives only in XML.

My issue is that your PR takes us further from this goal by adding a brand new design pattern, rather than working with the default/model/top level layout that we already have for the majority of scoreboard state.

I'm not asking you to go all the way and hook this into XML/autosaves/WS (I think we can get away without it for this specific feature), but I do ask you make to this look like Teams, Skaters and Positions do.

> while we are less than 2 weeks from the first continental cup.

We already ran out of time there. The build I've recommended to WFTDA is at the clock sync fix.

In relation to #48, I'm not sure why that's considered urgent. We've survived 9 years without that corner case in the rules being handled, and the traditional workaround everyone has (presumably) been using for years still works.

The clock weirdness I do care about, but any changes to that are high-risk and we simply don't have time to test.

Reply:
The thing is: Your vision of how the code should be structured is at this point just one developers vision. There has never been a discussion in the dev group or anywhere else that I am aware of on this topic.
If you want people to follow that vision, write it up and propose it in the Dev group, so we can have a proper discussion about it with all interested parties. Hopefully we will reach a consensus there that we can then point to in future PRs.
And I propose that this PR is merged, so development can continue. It's not as if adapting the snapshots to any agreed upon code structure will be made significantly harder by this step.

Reply:
You're proposing a new design pattern where by we start spreading state around tightly-coupled but distinct packages. As you are the one proposing a change the onus is on you gain support for it, not on me.

> It's not as if adapting the snapshots to any agreed upon code structure will be made significantly harder by this step.

Your proposed PR makes the code harder to understand and maintain in my opinion, and it is ultimately one more special case that'll need to be made not special later on. I'd much prefer we got it right first time around, especially as the previous iteration was quite close and this is a relatively small and self-contained change.

If you're not willing to change the code to be in line with how the existing code is structured, in order to resolve this I'm willing to build off your commits and do it for you.

Reply:
I do believe that your proposed way is harder to maintain. So I am not willing to implement that.